### ISSUE TITLE
Test request data population (body & query string)

### ISSUE THREAD

**Issue by noahdietz** [apigee-127/swagger-test-templates#96]
I am beginning work in my own fork on creating options for specifying the values of generated test request parameters in the body and query string, following a similar style to #91.

The options will take the form of JSON objects with values provided on a path-by-path, operation-by-operation basis. See the example below.

For a path specified as:
```json
"/user": {
  "post": {
    "parameters": [
      {
        "name": "name",
        "in": "body",
        "description": "name of a new User",
        "required": true,
        "type": "string"
      }
     ]
    }
  }
```
the corresponding parameter data would be passed to the function call options as:
```json
"bodyParams": {
  "/user": {
    "post": {
      "name": "John Doe"
    }
  }
}
```

**Comment by noahdietz** [comment]
Now have data population for:
* path params (from a separate branch of work)
* query string params
* body params

Will submit a PR from my fork.

**Comment by mm-gmbd** [comment]
Hey @noahdietz -- I'm coming back around to using STT and am in need of the exact same feature.  Any chance this PR is ready, or can I jump in to help complete?

**Comment by mm-gmbd** [comment]
Actually, I just glanced at your branch and saw your implementation of adding optional parameters `bodyParams` and `queryParams`.

A couple of notes:

- Is there a way to provide different sets of inputs for requests that should succeed vs. fail? It looks like they are only definable by route (rather than by route *and* success/failure)
- Using tools like [json-schema-faker](https://github.com/json-schema-faker/json-schema-faker), STT could automatically generate both pass/fail response bodies. This is more work, but definitely the version I'd like to see.

**Comment by noahdietz** [comment]
@mm-gmbd thanks for revisiting the project

* no, there is no way to differentiate data between passing & failing, although that would be much more desirable...I did this in my own fork as a sort of proof of concept for a paper i was writing.
*  json-schema-faker looks pretty cool to me, how do you propose specifying successful/failure? An idea that just came to me was you use a handlebars helper, that takes the expected status code (2XX for success, everything else failure) and the validation schema of the request, then based on the status code, either the schema is used to generate a successful body, or ignore the schema and generate a generally failing body (unless you want to use the schema and fudge it, but thatd be more work)...thoughts?

**Comment by noahdietz** [comment]
I would like to avoid changing the traversal hierarchy if possible...my suggestion:

We take the syntactic approach you've suggested in the STT config, allowing specification of values based on status code, stored in an array keyed to the path...._however_ we do the evaluation of the request parameters (i.e. a look up in the `bodyVals`, based on the status code && path being evaluated in the template) in the **handlebar helper function**. This will also be a lot less work, less diffs and an easy PR.

I know I keep suggesting the handlebars helper, but they are the most configurable aspect of the module. The traversal should stay the same, but the information being passed on to the Handlebars compilers can be changed (with changes to the templates too)

Does this make sense? What do you think @mm-gmbd

**Comment by mm-gmbd** [comment]
Hmm, I understand what you're saying, but I'm not sure how to realize the goal of generation of multiple test cases *per statuscode* in the current structure, unless the templates were more heavily modified.

And actually, maybe it's not as impossible as I'm thinking...

For example, let's look at `/templates/supertest/put/put.handebars`. Right now, it generates a single test, starting as follows:

```
  it('should respond with {{length description}}', function(done) {
      {{#validateResponse returnType noSchema}}
      /*eslint-disable*/
      {{> schema-partial this}}

      /*eslint-enable*/
      {{/validateResponse}}
      api.put({{pathify path pathParams}})
      ...
      {{#is contentType 'application/json'}}
      .send({
        {{#each bodyParameters}}
        {{this.name}}: {{bodify ../../path this.name ../../bodyVals}}{{#unless @last}},{{/unless}}
        {{/each}}
      })
      {{/is}}
      ...
  })
```

Rather than passing bodyVals through the `bodify` helper, the template could be modified to immediately loop through the `config.inputTesting` for that particular path, something like:

1. Define `inputTesting` like the following:

```
config.inputTesting["a/path"] = {
  "200": [
    {
      "bodyVals": //good body
      "pathVals": //good path
      "queryVals": //good query
    }
  ],
  "400": [
    {
      "bodyVals": //bad body 1
    },
    {
      "bodyVals": //bad body 2
    },
  ]
}
```

2. Modify index.js to get the inputTesting parameters for that particular operation/statusCode combination
```
if (config.inputTesting && config.inputTesting[path] && config.inputTesting[path][response]) {
    data.inputTesting = config.inputTesting[path][response];
  }
```

3. Modify template to look like the following:

```
{{#if inputTesting}} //see if inputTesting is defined for this particular operation/statusCode combination
  {{#each inputTesting}}
    //Do standard template, but replace the request body with the provided bodyVals by passing them to `bodify`
  {{/each}}
{{#else}}
  //Do the standard template, and indicate that the user needs to fill out the request body
{{/if}
```

**Comment by Remco75** [comment]
Hi all, any progress on this? I am currently in the process of implementing the some kinda functionality using the swagMock module for data generation. Will push it soon and make a PR if anyones interested that is ;-)

**Comment by mm-gmbd** [comment]
I have implemented a version of this but haven't submitted a PR, and I think the version I ended up with is closer to the way @noahdietz was hoping it would be.

Rather than the way I wrote it in my last comment (modifying the template to include a big `if/else`), I *slightly* altered the codegen (but I think it still fits the current paradigm) to ask for user input or utilzie the input testing if provided for that path/method.

I can probably submit a PR in the near future (today/tomorrow).
